src/ex2/CompileVisitor.java

```diff
@@ -2,63 +2,180 @@
         this.classMapping = classMapping;
     }
+
     public String getString() {
         return builder.toString();
     }
 
     @Override
     public void visit(Program program) {
+        CompileClassesVisitor ccv = new CompileClassesVisitor();
+        program.accept(ccv);
+        builder.append(ccv.getString());
+        symbolMapping = ccv.getMapping();
+        pointerMap = ccv.getPointerMap();
 
+        program.mainClass().accept(this);
+        for (var classDecl : program.classDecls())
+            classDecl.accept(this);
     }
 
 
     @Override
     public void visit(MainClass mainClass) {
-
     }
 
     // OOP
-    @Override
-    public void visit(ClassDecl classDecl) {
-    }
-
-    @Override
-    public void visit(MethodDecl methodDecl) {
 
+    private void addLine(String line) {
+        builder.append("\t").append(line).append("\n");
     }
 
     @Override
-    public void visit(FormalArg formalArg) {
-
+    public void visit(ClassDecl classDecl) {
+        for (var methodDecl : classDecl.methoddecls()) {
+            String s = String.format("define %s @%s.%s(i8* %%this",
+                    TypeDecider.llvmType(methodDecl.returnType()), classDecl.name(), methodDecl.name());
+            builder.append(s);
+            methodDecl.accept(this);
+        }
     }
 
     @Override
-    public void visit(VarDecl varDecl) {
-        // same as formal arg ??
+    public void visit(MethodDecl methodDecl) { // this compiles from the formals after %this
+        for (var formal : methodDecl.formals()) {
+            String s = String.format(", %s %%.%s", TypeDecider.llvmType(formal.type()), formal.name());
+            builder.append(s);
+        }
+        builder.append(") {\n");
+        lastRegisterNumber = 0; // it's that easy
+        for (var formal : methodDecl.formals())
+            formal.accept(this);
+        for (var varDecl : methodDecl.vardecls())
+            varDecl.accept(this);
+        for (var statement : methodDecl.body())
+            statement.accept(this);
+        builder.append("}\n\n");
     }
 
     @Override
     public void visit(MethodCallExpr e) {
-
+        // first determine what function is called
+        StaticClassVisitor classFinder = new StaticClassVisitor();
+        e.ownerExpr().accept(classFinder);
+        String ownerClass = classFinder.getResult();
+        SymbolTable st;
+        if (ownerClass.equals("current"))
+            st = e.getSymbolTable();
+        else
+            st = classMapping.get(ownerClass);
+        Symbol funcSymbol = st.methodLookup(e.methodId());
+
+        e.ownerExpr().accept(this); // this should generate the owner ptr
+        int ownerPtr = lastRegisterNumber;
+        List<Integer> actualsRegisters = new LinkedList<>();
+        for (var actual : e.actuals()) { // generate code for actuals and save their registers
+            actual.accept(this);
+            actualsRegisters.add(lastRegisterNumber);
```

Comment:
Shouldn't it be lastRegisterNumber - 1? Since you increment it each time you use a register